src/gc-alloc-profiler.cpp

```diff
@@ -0,0 +1,167 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#include "gc-alloc-profiler.h"
+
+#include "julia_internal.h"
+#include "gc.h"
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+using std::unordered_map;
+using std::string;
+using std::vector;
+
+struct RawBacktrace {
+    jl_bt_element_t *data;
+    size_t size;
+};
+
+struct RawAlloc {
+    jl_datatype_t *type_address;
+    RawBacktrace backtrace;
+    size_t size;
+};
+
+struct PerThreadAllocProfile {
+    vector<RawAlloc> allocs;
+    unordered_map<size_t, size_t> type_address_by_value_address;
+    unordered_map<size_t, size_t> frees_by_type_address;
+};
+
+struct AllocProfile {
+    double sample_rate;
+
+    vector<PerThreadAllocProfile> per_thread_profiles;
+};
+
+struct CombinedResults {
+    vector<RawAlloc> combined_allocs;
+    vector<FreeInfo> combined_frees;
+};
+
+// == global variables manipulated by callbacks ==
+
+AllocProfile g_alloc_profile;
+int g_alloc_profile_enabled = false;
+CombinedResults g_combined_results; // will live forever
+
+// === stack stuff ===
+
+RawBacktrace get_raw_backtrace() {
+    static jl_bt_element_t static_bt_data[JL_MAX_BT_SIZE];
+
+    size_t bt_size = rec_backtrace(static_bt_data, JL_MAX_BT_SIZE, 2);
+
+    size_t bt_bytes = bt_size * sizeof(jl_bt_element_t);
+    jl_bt_element_t *bt_data = (jl_bt_element_t*) malloc(bt_bytes);
+    memcpy(bt_data, static_bt_data, bt_bytes);
+
+    return RawBacktrace{
+        bt_data,
+        bt_size
+    };
+}
+
+// == exported interface ==
+
+JL_DLLEXPORT void jl_start_alloc_profile(double sample_rate) {
+    g_alloc_profile = AllocProfile{sample_rate};
+
+    for (int i = 0; i < jl_n_threads; i++) {
+        g_alloc_profile.per_thread_profiles.push_back(PerThreadAllocProfile{});
+    }
+
+    g_alloc_profile_enabled = true;
+}
+
+extern "C" {  // Needed since the function doesn't take any arguments.
+
+JL_DLLEXPORT struct RawAllocResults jl_fetch_alloc_profile() {
+    // TODO: check that the results exist
```

Comment:
maybe we should also check that the profile isn't still running… (i.e. you've called `stop`, which populates these results)

Reply:
😮 oh, good call... 🤔 i wonder what happens if you violate this in the CPU Profile?
```julia
julia> @profile for _ in 1:100000; Profile.clear(); Profile.fetch() ;end
```
🤔 doesn't seem to cause any problems... but _definitely_ weird! haha

Yeah, i think asserting that seems reasonable to me, too 👍 

It actually makes me wonder if maybe we want _both_ a (cheap) interface that simply dumps the stuff from C to Julia (equivalent of fetch) _and_ a separate decoding step that converts things to useful data? I'm not sure why the Profile.jl stuff separates those concepts... 🤔 I think what you have now is good though, and no reason to change it unless someone else comments on it.

Reply:
Okay we talked on a call, and we don't think there's anything that needs to be done here, since the scratch space to record profile is different from the combined results object where it's stored, so fetch() can't interleave badly with a running profile. :)